website_product_subscription: Do not re-create web access during validation

carmenbianca opened the pull request:
When a subscriber already has web access, don't run create_web_access
again during validation.

Previously, when this was done on requests of type 'gift', the
subscriber would have received a second e-mail.

This behaviour was first introduced in: https://github.com/coopiteasy/product-subscription/commit/a79510803f502b375fb01e9260cc79b079bbfefd

Task: https://gestion.coopiteasy.be/web#id=6534&view_type=form&model=project.task&action=479

carmenbianca committed:
[FIX] website_product_subscription: Do not re-create web access during validation

When a subscriber already has web access, don't run create_web_access
again during validation.

Previously, when this was done on requests of type 'gift', the
subscriber would have received a second e-mail.

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca commented on website_product_subscription/models/subscription_request.py:
```diff
@@ -11,7 +11,9 @@ class SubscriptionRequest(models.Model):
         res = super(SubscriptionRequest, self).validate_request()
-        self.create_web_access()
+        for request in self:
+            if not self.env["res.users"].user_exists(request.subscriber.email):
```
This line is duplicated from `website_product_subscription.ResPartner.create_web_access`:

```python
    @api.model
    def create_web_access(self):
        User = self.env["res.users"]
        for partner in self:
            if not User.user_exists(partner.email):
                User.create_user(
                    {"login": partner.email, "partner_id": partner.id}
                )
```

Ideally I want to instead create and use a method `_has_web_access`, but I'm a little confused that the above method is `@api.model`.

Would this patch in `res_partner.py` work?

```patch
diff --git a/website_product_subscription/models/res_partner.py b/website_product_subscription/models/res_partner.py
index 4e4cedb..309f84b 100644
--- a/website_product_subscription/models/res_partner.py
+++ b/website_product_subscription/models/res_partner.py
@@ -8,11 +8,16 @@ from openerp import models, api
 class ResPartner(models.Model):
     _inherit = "res.partner"
 
-    @api.model
+    @api.multi
+    def _has_web_access(self):
+        self.ensure_one()
+        return self.env["res.users"].user_exists(self.email)
+
+    @api.multi
     def create_web_access(self):
         User = self.env["res.users"]
         for partner in self:
-            if not User.user_exists(partner.email):
-                User.create_user(
+            if not partner._has_web_access():
+                self.env["res.users"].create_user(
                     {"login": partner.email, "partner_id": partner.id}
                 )
```

robinkeunen replied:
looking at where `ResPartner.create_web_access` is used, I do not see a reason why it should be tagged as `@api.model`. Your patch seems a good one to me. But will it fix the the issue ?

I also find it ugly that `def send_gift_emails` creates the web access, it does more than its name suggests. Could you put the `ResPartner.create_web_access` call back to `SubscriptionRequest.create_web_access`?

carmenbianca replied:
The original PR fixes the issue. The patch in the comments is just niceties.

I also didn't like that send_gift_emails creates web access, but I wasn't sure how stable our APIs should be, so I worked around it. I can refactor that.

carmenbianca committed:
[REF] website_product_subscription: Uncouple web access creation from send_gift_emails

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca commented:
Refactored and tested. Only thing I haven't tested is the cron job.

carmenbianca commented:
This patch may need to be committed later. I'm anxious that the test server will start sending a lot of backlog e-mails once I commit this.

This patch makes sure that the cronjob (cron_create_scheduled_gift_user) doesn't crash halfway through when it attempts to create a user without email address. I created the check in two places, but it would technically suffice to have it only in `res_partner.py`.

Not sure whether logging an error is sufficient error handling. Should it alert an admin instead?

This intersects with https://gestion.coopiteasy.be/web#id=6535&view_type=form&model=project.task&menu_id=338&action=479 which causes the problem of requests-without-email-addresses existing in the first place.

```patch
diff --git a/website_product_subscription/data/cron.xml b/website_product_subscription/data/cron.xml
index 9fbc85d..4b65152 100644
--- a/website_product_subscription/data/cron.xml
+++ b/website_product_subscription/data/cron.xml
@@ -9,10 +9,9 @@
         <field name="interval_number">24</field>
         <field name="interval_type">hours</field>
         <field name="numbercall">-1</field>
-        <field name="doall" eval="False"/>
+        <field name="doall" eval="True"/>
         <field name="model">product.subscription.request</field>
         <field name="function">cron_create_scheduled_gift_user</field>
         <field name="args">()</field>
     </record>
 </odoo>
-
diff --git a/website_product_subscription/models/res_partner.py b/website_product_subscription/models/res_partner.py
index 204a704..394ad25 100644
--- a/website_product_subscription/models/res_partner.py
+++ b/website_product_subscription/models/res_partner.py
@@ -2,8 +2,12 @@
 #   Robin Keunen <robin@coopiteasy.be>
 # License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
 
+import logging
+
 from openerp import models, api
 
+_logger = logging.getLogger(__name__)
+
 
 class ResPartner(models.Model):
     _inherit = "res.partner"
@@ -16,6 +20,11 @@ class ResPartner(models.Model):
     @api.multi
     def create_web_access(self):
         for partner in self:
+            if not partner.email:
+                _logger.error(
+                    "%s does not have an email address; cannot create web access"
+                )
+                continue
             if not partner._has_web_access():
                 self.env["res.users"].create_user(
                     {"login": partner.email, "partner_id": partner.id}
diff --git a/website_product_subscription/models/subscription_request.py b/website_product_subscription/models/subscription_request.py
index 0a1451a..b12947a 100644
--- a/website_product_subscription/models/subscription_request.py
+++ b/website_product_subscription/models/subscription_request.py
@@ -2,8 +2,12 @@
 #   Robin Keunen <robin@coopiteasy.be>
 # License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
 
+import logging
+
 from openerp import models, fields, api
 
+_logger = logging.getLogger(__name__)
+
 
 class SubscriptionRequest(models.Model):
     _inherit = "product.subscription.request"
@@ -38,6 +42,11 @@ class SubscriptionRequest(models.Model):
     def create_web_access(self):
         today = fields.Date.today()
         for request in self:
+            if not request.subscriber.email:
+                _logger.error(
+                    "%s does not have an email address; cannot create web access"
+                )
+                continue
             if request.type == "gift" and request.gift_date <= today:
                 request.send_gift_emails()
             elif request.type == "gift" and request.gift_date > today:
```

robinkeunen commented on website_product_subscription/models/subscription_request.py:
```diff
@@ -25,11 +27,11 @@ def send_gift_emails(self):
+            template.send_mail(request.id)
             request.gift_sent = True
 
     @api.multi
```
Could you refactor this function to `def create_subscriber_web_access` ?

I'm sorry, I could have sait it in my first review but I got confused twice :/

carmenbianca replied:
+1

robinkeunen commented on website_product_subscription/models/subscription_request.py:
```diff
@@ -53,4 +55,4 @@ def cron_create_scheduled_gift_user(self):
             ]
         )
-        requests.send_gift_emails()
+        requests.create_web_access()
```
Bien vu 👍

carmenbianca replied:
:)

robinkeunen replied:
You refactored that function name to `create_subscriber_web_access`

robinkeunen replied:
But I also suggested to remove that function altogether in another comment.

robinkeunen commented on website_product_subscription/models/subscription_request.py:
```diff
@@ -11,7 +11,9 @@ class SubscriptionRequest(models.Model):
     def validate_request(self):
         res = super(SubscriptionRequest, self).validate_request()
-        self.create_web_access()
+        for request in self:
+            if not request.subscriber._has_web_access():
+                request.create_web_access()
```
I think you should remove this test: if the user already has a web access (from current subscription) and is gifted a new subscription, he will not receive the _existing user gift email_.

carmenbianca replied:
This is the code that fixes the problem, though.

On the wilfried DB, `validate_request()` is called by Odoo a few seconds after `process_form()` (controllers/subscribe.py) is finished\*. It goes through `create[_subscriber]_web_access()` to `send_gift_emails()`, neither of which have any check along the lines of 'if an email was already sent, don't send another one'. In this way, `create[_subscriber]_web_access` was called twice: once from `process_form()`, and once from `validate_request()`. I could (and probably should) add that check in either/both functions like so:

```python
for request in self:
    if request.gift_sent:
        continue
```

But even if I do that, the check you highlighted should probably remain. The program shouldn't create access for users that already have it, even if the `create_web_access()` function is highly sanitised to skip things that have already been executed.

So if we remove the check (and add sanitation against `request.gift_sent`), the program is going to needlessly call `create_web_access()` a second time, and do essentially nothing.

---

\* It didn't do this on a virgin DB, which cost me quite a bit of time in reproducing this issue.

robinkeunen commented:
And about your questions:

**emails:** do not worry about backlog emails from the test server, we set dummy values to the mail server configuration when restoring a db with ociedoo. cf `/home/odoo9/.config/ociedoo/posthook.sql`.

**doall**: it is not needed here since the cron processes all gift requests where `("gift_date", "<=", today),`.

**logging an error** it's probably not enough to log an error, maybe ask an analyst whether you should send a notification email.

carmenbianca committed:
[FIX] website_product_subscription: Make sure cron_create_scheduled_gift_user doesn't crash

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca committed:
[IMP] website_product_subscription: Run cron_create_scheduled_gift_user ONLY for today's queue

This makes sure that messages that should have been sent yesterday (or
worse: weeks ago) aren't sent today (or: ages later).

To make this a little more resilient, I've upped the amount of times the
cronjob is run. I've also disabled the cronjob during sleeping hours.

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca commented:
@robinkeunen Done with refactoring. This needs a final review. Merci :)

robinkeunen commented on website_product_subscription/models/res_partner.py:
```diff
@@ -8,11 +8,15 @@
 
-    @api.model
+    @api.multi
+    def _has_web_access(self):
```
Since the method `_has_web_access` is accessed outside of res_partner, the leading underscore should fall.

carmenbianca replied:
+1

robinkeunen commented on website_product_subscription/controllers/subscribe.py:
```diff
@@ -100,7 +100,8 @@ def process_form(self):
         sub_req = self.create_subscription_request()
-        sub_req.create_web_access()
+        sub_req.send_gift_emails()
+        sub_req.create_subscriber_web_access()
```
Since you emptied the body of this function in your much needed rafactoring, I would remove that function altogether and replace it with : `sub_req.mapped("subscriber").create_web_access()`

robinkeunen replied:
This is a matter of taste, feel free to leave it as is.

carmenbianca replied:
Sure, can adjust.

carmenbianca commented on website_product_subscription/models/subscription_request.py:
```diff
@@ -25,32 +34,57 @@ def send_gift_emails(self):
+            return
+
         requests = self.search(
             [
                 ("gift", "=", True),
                 ("gift_sent", "=", False),
-                ("gift_date", "<=", today),
+                # TODO: Revert following back to '<=' once the back log has
+                # cleared. (T3833)
+                ("gift_date", "=", fields.Date.today()),
             ]
         )
```
Should the following be added?

```python
("state", "=", "paid")
```

or something like

```python
("state", "!=", "cancel")
```

The states are `["draft", "sent", "paid", "cancel"]`

robinkeunen replied:
This is a question for Wilfried or Medor, I'd say `("state", "in", ("sent", "paid"))` because some people pay with a later bank transfert.

carmenbianca replied:
I can't really imagine any scenario where cancelled requests should receive e-emails. Asked Cath about this.

robinkeunen replied:
sure but I'm not so sure about "draft" requests.

carmenbianca replied:
Talked with Cath; we'll keep this as-is and cheat by setting `gift_sent` to True for people in the back log who will not receive e-mails.

robinkeunen commented on website_product_subscription/models/subscription_request.py:
```diff
@@ -60,12 +61,22 @@ def create_subscriber_web_access(self):
     @api.model
     def cron_create_scheduled_gift_user(self):
-        today = fields.Date.today()
+        now = fields.Datetime.from_string(fields.Datetime.now()).time()
+        if now < datetime.time(7, 0, 0) or now > datetime.time(22, 0, 0):
+            # Don't send e-mails during hours meant for sleep.
+            _logger.info(
+                "Skipping cron_create_scheduled_gift_user because it's late: %s"
+                % now
+            )
+            return
```
Was that a request from wilfried ?

With this code + the condition `("gift_date", "=", fields.Date.today()),`, gift requests

-  submitted today after 22:00
- with gift_date = today

Will never be sent.

robinkeunen replied:
Your best bet would probably be `gift_date between now - timedelta(hours=24) and now`

carmenbianca replied:
:thinking: you're right. It wasn't a request from Wilfried, but in decreasing the interval in the cron, I figured that it'd be courteous to do this.

Following options:

1. Decrease the interval back to once per day and hope that that's sufficient.
2. Remove the above check and accept that e-mails are going to be sent at 01:00 at night.
3. Keep everything as-is, but change the condition back to `("gift_date", "<=", today)`, and sanitise the DB (i.e., disable/cancel/delete gift requests that are too old) before pushing this to prod.

robinkeunen replied:
I think you should remove the night tests indeed: KISS. Did you see my time window suggestion?

carmenbianca replied:
Did not see time window suggestion. Was it in the cron video?

robinkeunen replied:
In my comment just above : `gift_date between now - timedelta(hours=24) and now`

carmenbianca replied:
Not sure how to do that (is the above code or pseudocode?), but it's probably no longer necessary. I've reverted this commit to sending e-mails once per day, including gifts from the past.

Cath will be able clear the backlog before we push to prod, so the workaround is not needed.

carmenbianca committed:
Revert "[IMP] website_product_subscription: Run cron_create_scheduled_gift_user ONLY for today's queue"

This reverts commit 249ae41613e5ca9fdb0b89b4c184300b09fb17f9.

robinkeunen reviewed (approved):
Much clearer now, thank you 👌
